### ISSUE TITLE
[16] MarkupLM: Pre-training of Text and Markup Language for Visually-rich Document Understanding

### ISSUE THREAD

**Issue by dhkim0225** [dhkim0225/1day_1paper#43]
LayoutLM 저자들의 다음 행선지.

[paper](https://arxiv.org/pdf/2110.08518.pdf)
[code](https://github.com/microsoft/unilm/tree/master/markuplm)

![스크린샷, 2021-10-25 11-16-26](https://user-images.githubusercontent.com/16400591/138629119-4ef461fa-a6c2-419c-b199-5bec0500ab91.png)

## XPath Embedding
html 페이지를 갖고 pretraining 을 하기 위해서 xpath 를 이용한다.
[xpath 블로그](https://loco-motive.tistory.com/26) 의 `1.3.3 predicates` 부분을 보면 이해에 도움이 된다.

i 번째 input token 에 대해서 xpath expression을 다음과 같이 list로 우선 나타낸다.
논문의 예시를 보면 하나의 node(tree 의 한 가로줄 영역) 을 xpath로 나타내고 `/html/body/div/li[1]/div/span[2]` 요걸 리스트로 나타낸 거다.
![image](https://user-images.githubusercontent.com/16400591/138630421-7c2d695e-a37d-49a9-baa6-05dca78a81f4.png)
여기서 t는 tag_name, s 는 subscript 이다. subscript 가 없다면 0으로 넣어준다고 한다.

tag_name, subscript 각각에 embedding 취해주면 `i 번째 token의 j번째 xpath expression 에 대한 embedding` 이 완성된다.
tag embedding 크기는 216, subscript embedding 크기는 1001
tag, subscript embedding dimension 모두 32
![image](https://user-images.githubusercontent.com/16400591/138630543-a7ebcf63-1e02-477c-bf86-544cff18391a.png)

요 녀석을 concat 해주고, linear project 를 해주면 xpath embedding 완성이다.
![image](https://user-images.githubusercontent.com/16400591/138630890-6377b90d-8a52-47e1-8499-53a2db2c0a1b.png)
![image](https://user-images.githubusercontent.com/16400591/138630916-0fb10336-c2f6-43e9-a781-a796381ea474.png)

요 과정을 이미지 하나에 담아보면, 요렇다.
![image](https://user-images.githubusercontent.com/16400591/138630955-ddbc19b4-19a3-4ef8-a61d-f2b4b773807a.png)

Xpath embedding 수행 시 max depth는 50으로 두었다고 한다.

## Pre-training Objectives
3가지 task 를 수행한다. MMLM, NRP, TPM
![image](https://user-images.githubusercontent.com/16400591/138631149-5472b1f2-50b1-4dbc-a68c-7213add52db8.png)
![image](https://user-images.githubusercontent.com/16400591/138631733-77bbc4bd-79a2-4fb8-9fe4-b908a553a9b2.png)

#### MMLM (Masked Markup Language Modeling)
랜덤하게 token masking 15%, webpage 제목 쪽은 TPM을 위해서 masking 하지 않았음.

#### NRP (Node Relation Prediction)
`R ∈ {self, parent, child, sibling, ancestor, descendent, others}` 각각의 node 마다 7개의 관계를 할당해 두고 node 의 첫번째 token 들끼리의 관계를 예측하도록 한다. 

#### TPM (Title-Page Matching)
15% 확률로 title 쪽 제목을 랜덤 token으로 바꿔버리고, [CLS] 에 linear 붙여서 `title 바뀌었나, 안 바뀌었나 binary classification` 수행

## Results
![image](https://user-images.githubusercontent.com/16400591/138631920-6e5f5a64-80a5-4473-836e-8d5500b29b8a.png)
![image](https://user-images.githubusercontent.com/16400591/138631969-bd45fc3e-cc49-4fc9-a161-91e0eb655646.png)
